### PR TITLE
Add option compose-node-name

### PR DESCRIPTION
**Pull request by gburiola**
This allows node files in different subfolders to have the same name

More information on README-extentions.rst

### CODE REVIEW

File: reclass/config.py [hunk 1]
```diff
@@ -32,6 +32,9 @@ def make_db_options_group(parser, defaults={}):
     ret.add_option('-z', '--ignore-class-notfound', dest='ignore_class_notfound',
                    default=defaults.get('ignore_class_notfound', OPT_IGNORE_CLASS_NOTFOUND),
                    help='decision for not found classes [%default]')
+    ret.add_option('-a', '--add-subdir-to-node', dest='add_subdir_to_node', action="store_true",
+                   default=defaults.get('full_node_path', OPT_ADD_SUBDIR_TO_NODE),
+                   help='Add subdir when generating node names. [%default]')
```
Comment (gburiola) [inline reclass/config.py]: I'm happy to change the long name `add_subdir_to_node`, short name or description to anything else.

File: reclass/storage/yaml_fs/__init__.py [hunk 2]
```diff
@@ -50,12 +50,16 @@ def _path_mangler_inner(path):
...
         super(ExternalNodeStorage, self).__init__(STORAGE_NAME)
 
         if nodes_uri is not None:
             self._nodes_uri = nodes_uri
-            self._nodes = self._enumerate_inventory(nodes_uri, NameMangler.nodes)
+            if add_subdir_to_node:
+                self._nodes = self._enumerate_inventory(nodes_uri, NameMangler.classes)
+            else:
+                self._nodes = self._enumerate_inventory(nodes_uri, NameMangler.nodes)
+
```
Comment (gburiola) [inline reclass/storage/yaml_fs/__init__.py]: This is the main change of this MR

if `add_subdir_to_node` is true, then we use the same `NameMangler` used for classes that concatenates all subdirs with `.` and add filename at the end:

example:

```
inventory/nodes/prod/mysql.yaml
```

returns

```
prod.mysql
```

**Comment by gburiola** [comment]
Hey @epcim / @AndrewPickford 

Do you have any feedback on this MR ?

**Comment by AndrewPickford** [comment]
The basic idea looks fine, not something I'll use but that covers a fair few other options.

For the patch maybe make the add_subdir_to_node parameter mandatory in the reclass/__init__.py get_storage method. That way there's no need to import OPT_ADD_SUBDIR_TO_NODE for defaults in reclass/__init__.py and the default option is set in the settings class.

As well as the cli class the two adapter classes will also need changing so that add_subdir_to_node option is correctly passed into reclass. Otherwise running reclass stand alone would give a different answer than running via the salt/ancible adapters.

For the settings class the add_subdir_to_node option should also be tested in the __eq__ method. Also the option should also be used consistently by both the yaml_fs and yaml_git storage classes. That maybe something I'll need to look at.

If you're happy with these suggestions then update your fork and once that's ready I'll make a new branch in this repo for this work and merge in the changes. Then we can polish things up before merging into develop.

**Comment by epcim** [comment]
I am not the user of the feature as well. I woud ask to rename it, different then what it does vs. wha it provides.Also an concept usage ande better example in readme?One question: would that work even with other backends than yaml_fs?We rather use formula "reclass" that recently got class-maping features and dynamic node registration. In my usecase all files under nodes (except 5line salt master) are generated by reclass.storage.node state.PMiOn 26 Jun 2018 14:57, Andrew Pickford <notifications@github.com> wrote:The basic idea looks fine, not something I'll use but that covers a fair few other options.
For the patch maybe make the add_subdir_to_node parameter mandatory in the reclass/init.py get_storage method. That way there's no need to import OPT_ADD_SUBDIR_TO_NODE for defaults in reclass/init.py and the default option is set in the settings class.
As well as the cli class the two adapter classes will also need changing so that add_subdir_to_node option is correctly passed into reclass. Otherwise running reclass stand alone would give a different answer than running via the salt/ancible adapters.
For the settings class the add_subdir_to_node option should also be tested in the eq method. Also the option should also be used consistently by both the yaml_fs and yaml_git storage classes. That maybe something I'll need to look at.
If you're happy with these suggestions then update your fork and once that's ready I'll make a new branch in this repo for this work and merge in the changes. Then we can polish things up before merging into develop.

**Comment by epcim** [comment]
Any update?

**Comment by gburiola** [comment]
Thanks for your replies @epcim and @AndrewPickford. I was quite busy in the last couple of weeks.
I'll look at your suggestions on the weekend.

**Comment by epcim** [comment]
@gburiola possibly something  bit "shorter", as it's used from the command line

(some ideas, not the best, it's tough :) ):
```
node_name_from_path
node_name_composite
compose_node_name
composite_minion_id
use_path_in_minion_id
minion_id_from_path                   <-- what about this?
```
Finally, I think you should ignore any `_` prefixed directories, as for example, we use`_generated` as we generate nodes by reclass formula. Others might have similar case.

**Comment by gburiola** [comment]
@epcim thanks for the feedback.
for the flag name, I prefer to use `node` instead of `minion` because it's more generic and not specific to salt. Shall we go with `compose_node_name` ?

Regarding the `_generated` folder, not sure I understood. is this the behaviour you expect:

input (compose_node_name=true):

```
nodes/prod/mysql.yaml
nodes/_generated/server.yaml
```

output

```
prod.mysql
server
```

Can you send me an example of your node folder tree ?

**Comment by epcim** [comment]
i agree. compose node name is great. your example  is right. in _generared or whatever starting _ do not use compose_node_name. pmiOn 21 Jul 2018 23:07, Luis Buriola <notifications@github.com> wrote:@epcim thanks for the feedback.
for the flag name, I prefer to use node instead of minion because it's more generic and not specific to salt. Shall we go with compose_node_name ?
Regarding the _generated folder, not sure I understood. is this the behaviour you expect:
input (compose_node_name=true):
nodes/prod/mysql.yaml
nodes/_generated/s9upóerver.yaml

output
prod.mysql
server

Can you send me an example of your node folder tree ?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

1

File: reclass/config.py [hunk 3]
```diff
@@ -36,6 +36,9 @@ def make_db_options_group(parser, defaults={}):
     ret.add_option('-z', '--ignore-class-notfound', dest='ignore_class_notfound',
                    default=defaults.get('ignore_class_notfound', OPT_IGNORE_CLASS_NOTFOUND),
                    help='decision for not found classes [%default]')
+    ret.add_option('-a', '--add-subdir-to-node', dest='compose_node_name', action="store_true",
```
Comment (epcim) [inline reclass/config.py]: '-a', '--add-subdir-to-node' 
to
'-a', '--compose-node-name'

Comment (gburiola) [reply reclass/config.py]: thanks for spotting that. Fixed

**Comment by gburiola** [comment]
```
$ find /tmp/reclass -type f
/tmp/reclass/nodes/prod/mysql1.yml
/tmp/reclass/nodes/prod/_generated/generated3.yml
/tmp/reclass/nodes/staging/mysql2.yml
/tmp/reclass/nodes/_generated/prod/generated2.yml
/tmp/reclass/nodes/_generated/generated1.yml
```

```
$ python3 reclass.py  --inventory -b /tmp/reclass/ | yq  .nodes | jq 'keys'
[
  "generated1",
  "generated2",
  "generated3",
  "mysql1",
  "mysql2"
]
```

```
$ python3 reclass.py  --inventory --compose-node-name -b /tmp/reclass/ | yq  .nodes | jq 'keys'
[
  "generated1",
  "generated2",
  "prod._generated.generated3",
  "prod.mysql1",
  "staging.mysql2"
]
```

**Comment by gburiola** [comment]
Regarding your 3rd question above (having nodes/mysql.yml and nodes/prod/mysql.yml), they will NOT be merged. They will simply be treated as 2 separate nodes. One called `mysql` and the other called `prod.mysql`

File: reclass/storage/common.py [hunk 4]
```diff
@@ -13,6 +13,17 @@ def nodes(relpath, name):
         # no mangling required
         return relpath, name
 
+    @staticmethod
+    def composed_nodes(relpath, name):
+        if relpath == '.' or relpath == '':
+            # './' is converted to None
+            return None, name
+        parts = relpath.split(os.path.sep)
+        if parts[0].startswith("_"):
```
Comment (gburiola) [inline reclass/storage/common.py]: @epcim this is the code to ignore compose_node when path starts with `_`

Notice how it ONLY ignores `_` when it's at the beginning of the path

`nodes/_generated/mysql.yml` - subfolder is ignored
`nodes/prod/_generated/mysql.yml` - subfolder is NOT ignored

**Pull request merged**